src/core/lib/event_engine/posix_engine/posix_engine_listener_utils.cc

```diff
@@ -0,0 +1,380 @@
+      options.allow_reuse_port && socket.addr.address()->sa_family != AF_UNIX) {
+    GRPC_RETURN_IF_ERROR(socket.sock.SetSocketReusePort(1));
+  }
+
+#ifdef GRPC_LINUX_ERRQUEUE
+  if (!socket.sock.SetSocketZeroCopy().ok()) {
+    // it's not fatal, so just log it.
+    gpr_log(GPR_DEBUG, "Node does not support SO_ZEROCOPY, continuing.");
+  } else {
+    socket.zero_copy_enabled = true;
+  }
+#endif
+
+  GRPC_RETURN_IF_ERROR(socket.sock.SetSocketNonBlocking(1));
+  GRPC_RETURN_IF_ERROR(socket.sock.SetSocketCloexec(1));
+
+  if (socket.addr.address()->sa_family != AF_UNIX) {
+    GRPC_RETURN_IF_ERROR(socket.sock.SetSocketLowLatency(1));
+    GRPC_RETURN_IF_ERROR(socket.sock.SetSocketReuseAddr(1));
+    socket.sock.TrySetSocketTcpUserTimeout(options, false);
+  }
+  GRPC_RETURN_IF_ERROR(socket.sock.SetSocketNoSigpipeIfPossible());
+  GRPC_RETURN_IF_ERROR(socket.sock.ApplySocketMutatorInOptions(
+      GRPC_FD_SERVER_LISTENER_USAGE, options));
+
+  if (bind(fd, socket.addr.address(), socket.addr.size()) < 0) {
+    return absl::FailedPreconditionError(
+        absl::StrCat("Error in bind: ", std::strerror(errno)));
+  }
+
+  if (listen(fd, GetMaxAcceptQueueSize()) < 0) {
+    return absl::FailedPreconditionError(
+        absl::StrCat("Error in listen: ", std::strerror(errno)));
+  }
+  socklen_t len = static_cast<socklen_t>(sizeof(struct sockaddr_storage));
+
+  if (getsockname(fd, const_cast<sockaddr*>(sockname_temp.address()), &len) <
+      0) {
+    return absl::FailedPreconditionError(
+        absl::StrCat("Error in getsockname: ", std::strerror(errno)));
+  }
+
+  socket.port = SockaddrGetPort(ResolvedAddress(sockname_temp.address(), len));
+  // No errors. Set close_fd to false to ensure the socket is not closed.
+  close_fd = false;
+  return absl::OkStatus();
+}
+
+}  // namespace
+
+absl::StatusOr<ListenerSocket> CreateAndPrepareListenerSocket(
+    const PosixTcpOptions& options, const ResolvedAddress& addr) {
+  ResolvedAddress addr4_copy;
+  ListenerSocket socket;
+  auto result = PosixSocketWrapper::CreateDualStackSocket(
+      nullptr, addr, SOCK_STREAM, 0, socket.dsmode);
+  if (!result.ok()) {
+    return result.status();
+  }
+  socket.sock = *result;
+  if (socket.dsmode == PosixSocketWrapper::DSMODE_IPV4 &&
+      SockaddrIsV4Mapped(&addr, &addr4_copy)) {
+    socket.addr = addr4_copy;
+  } else {
+    socket.addr = addr;
+  }
+
+  GRPC_RETURN_IF_ERROR(PrepareSocket(options, socket));
+  GPR_ASSERT(socket.port > 0);
+  return socket;
+}
+
+absl::StatusOr<int> ListenerContainerAddAllLocalAddresses(
+    ListenerSocketsContainer& listener_sockets, const PosixTcpOptions& options,
+    int requested_port) {
+#ifdef GRPC_HAVE_IFADDRS
+  absl::Status status;
+  struct ifaddrs* ifa = nullptr;
+  struct ifaddrs* ifa_it;
+  bool no_local_addresses = true;
+  int assigned_port = 0;
+  if (requested_port == 0) {
+    auto result = GetUnusedPort();
+    if (!result.ok()) {
+      return result.status();
+    } else if (*result <= 0) {
```

Comment:
optional nit: the `else` is unnecessary since the previous line returns from the function. This applies to the `continue` lines below as well.

Reply:
Fixed these